master/internal/user/service.go

```diff
@@ -48,40 +48,74 @@ func (h *agentUserGroup) Validate() (*model.AgentUserGroup, error) {
 
 // Service describes a user manager.
 type Service struct {
-	db     *db.PgDB
-	system *actor.System
+	db        *db.PgDB
+	system    *actor.System
+	extConfig *model.ExternalSessions
 }
 
 // New creates a new user service.
-func New(db *db.PgDB, system *actor.System) (*Service, error) {
-	return &Service{db, system}, nil
+func New(db *db.PgDB, system *actor.System, extConfig *model.ExternalSessions) (*Service, error) {
+	return &Service{db, system, extConfig}, nil
 }
 
-// ProcessAuthentication is a middleware processing function that attempts
-// to authenticate incoming HTTP requests.  Note that the middleware looks
-// for an authentication in three places (in the following order):
+// The middleware looks for a token in two places (in this order):
 // 1. The HTTP Authorization header.
 // 2. A cookie named "auth".
-// 3. A Query parameter named "_auth".
+func (s *Service) extractToken(c echo.Context) (string, error) {
+	authRaw := c.Request().Header.Get("Authorization")
+	if authRaw != "" {
+		// We attempt to parse out the token, which should be
+		// transmitted as a Bearer authentication token.
+		if !strings.HasPrefix(authRaw, "Bearer ") {
+			return "", echo.ErrUnauthorized
+		}
+		return strings.TrimPrefix(authRaw, "Bearer "), nil
+	} else if cookie, err := c.Cookie("auth"); err == nil {
+		return cookie.Value, nil
+	}
+	// If we found no token, then abort the request with an HTTP 401.
+	return "", echo.NewHTTPError(http.StatusUnauthorized)
+}
+
+// ProcessAuthentication is a middleware processing function that attempts
+// to authenticate incoming HTTP requests.
 func (s *Service) ProcessAuthentication(next echo.HandlerFunc) echo.HandlerFunc {
```

Comment:
since this is auth logic, i really like it to be dedup'd as much as possible to avoid careless mistakes.. we should refactor ProcessAuthentication/ProcessExternalAuthentication/grpcutil.GetUser to avoid this probably, at some point. there was a lot you had to remember to do here.